crates/ruff/src/rules/dlint/rules/bad_shelve_use.rs

```diff
@@ -0,0 +1,42 @@
+use rustpython_parser::ast;
+use rustpython_parser::ast::{Ranged, Stmt};
+
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+
+use crate::checkers::ast::Checker;
+
+#[violation]
+pub struct BadShelveUse;
+
+impl Violation for BadShelveUse {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Use of the `shelve` module should be avoided")
+    }
+}
+
+/// DUO119
+pub(crate) fn bad_shelve_use(checker: &mut Checker, stmt: &Stmt) {
+    match stmt {
+        Stmt::Import(ast::StmtImport { names, range: _ }) => {
+            for name in names {
+                if &name.name == "shelve" {
+                    checker
+                        .diagnostics
+                        .push(Diagnostic::new(BadShelveUse, stmt.range()));
+                }
+            }
+        }
+        Stmt::ImportFrom(ast::StmtImportFrom {module, .. }, ..) => {
+            if let Some(id) = module {
+                if id == "shelve" {
+                    checker
+                        .diagnostics
+                        .push(Diagnostic::new(BadShelveUse, stmt.range()));
+                }
+            }
+        }
+        _ => panic!("Expected Stmt::Import | Stmt::ImportFrom"),
```

Comment:
You can now define your own enum to avoid having to use `panic`

```rust
#[derive(Debug, Copy, Clone)]
enum AnyStmtImport<'a> {
	Import(&'a ast::StmtImport),
	ImportFrom(&'a ast::StmtImportFrom)
}

impl AnyStmtImport<'_> {
	pub fn cast<'a>(stmt: &'a Stmt) -> Option<Self<'a>> {
		match stmt {
			Stmt::Import(import) => Some(Self::Import(import)),
			Stmt::ImportFrom(import_from) => Some(Self::ImportFrom(import_from)),
			_ => None,
		}
	}
}

impl<'a From<&'a ast::StmtImport> for AnyStmtImport<'a> {
	pub fn from(value: &'a ast::StmtImport) -> Self { Self::Import(value) }
}

impl<'a From<&'a ast::StmtImportFrom> for AnyStmtImport<'a> {
	pub fn from(value: &'a ast::StmtImportFrom) -> Self { Self::ImportFrom(value) }
}

pub(crate) fn bad_shelve_use(checker: &mut Checker, import: AnyStmtImport) {
```

Reply:
Thanks, gave it a go, let me know if I have implemented this correctly. Ran into some lifetime issues and did some trial-and-erroring to get it all to compile. Still fairly new to rust so happy for any feedback on implementation.